src/blog/0.3/linear-types-borrowing.vmd

```diff
@@ -0,0 +1,673 @@
+
+
+To get a sense of this "linear style", I like to compare it to Rust's borrow checking, because they're surprisingly similar.
+
+
+They both have some benefits:
+
+ * Memory safety with no direct extra costs. [# I say "direct" because it does have some indirect overhead; guideline #2 leads us to put more objects into arrays and hash maps, which leads to more bounds checking and hashing. Sometimes we also get more cache misses, though sometimes we get less.]
+ * Safety from data races; data races happen when two threads use references to access the same object, but we only ever allow one reference.
+
+
+And they both have some drawbacks:
+
+ 1. We often can't have multiple references to some data (except temporarily with borrow checking [# "Shared" references in Rust parlance, and they're read-only unless we use them to get to a `Cell`, `RefCell`, `Mutex`, etc.] [# Though, in both linear typing and borrow checking, long-lived references to the same data will still need to be made into indices or IDs into a central collection which holds the data.]).
+ 1. When we would otherwise have multiple references, we instead sometimes need to "slice" our data in unintuitive ways, which can cause some extra refactoring.
+ 1. We can't use optimal patterns like [intrusive data structures](https://lwn.net/Articles/907876/) and [graphs](https://news.ycombinator.com/item?id=24996001) which are sometimes faster in certain situations.
+ 1. We can't use certain patterns that decouple our code and make it more flexible, like [observers](https://www.reddit.com/r/rust/comments/pwqju6/is_there_an_underlying_reason_that_idiomatic_rust/), [back-references](https://users.rust-lang.org/t/back-reference-to-the-parent-struct/7413/2), [dependency references](https://en.wikipedia.org/wiki/Dependency_injection), [callbacks](https://news.ycombinator.com/item?id=12029238), and [delegates](https://medium.com/@nimjea/delegation-pattern-in-swift-4-2-f6aca61f4bf5).
+
+
+All of the above drawbacks exist because neither linear typing nor borrow checking can have *shared mutability*. In other words, multiple mutable references to the same data.
+
+
+<slice/>
+
+# Addressing the Drawbacks
+
+Are there any ways we can address those downsides without sacrificing the benefits?
+
+
+In other words, is there a way we can have linear types _and_ shared mutability?
+
+
+Adding borrow checking partially helps with drawback #1 [# Partially because borrow checking lets us have temporary references, not persistent ones. The borrow checker has a hard time putting references into data structures.] but it still has drawbacks #2, #3, and #4. Rust's `Rc` and `RefCell` helps with those, but they're incompatible with linear types' benefits [# We can't put a linear type into an `Rc` because `Rc` throws away its contents when the last alias disappears.], and have their own drawbacks (like forcing heap allocation and not having [higher RAII](https://verdagon.dev/blog/higher-raii-7drl)).
```

Comment:
"Rust's `Rc` and `RefCell` helps with those, but they're incompatible with linear types' benefits"

Maybe make this clearer that these things help Rust but won't help Vale. Like "Rust's RC and RefCell helps with those drawbacks, but we can't use those approaches in Vale. They are incompatible with linear types's benefits and additionally have their own drawbacks (etc...)